src/python/WMCore/MicroService/Unified/MSRuleCleaner.py

```diff
@@ -50,27 +121,337 @@ def execute(self, reqStatus):
+        self.logger.info("Workflows archived: %d", self.wfCounters['archived'])
+        return totalNumRequests, cleanNumRequests, archivedNumRequests
+
+    def _dispatchWflow(self, wflow):
+        """
+        A function intended to dispatch a workflow (e.g based on its status)
+        through one or more functional pipelines in case there is some more
+        complicated logic involved in the order we execute them but not just
+        a sequentially
+        """
+        # msg = "Called with wflow: %s"
+        # self.logger.debug(msg, pformat(wflow))
+        # NOTE: The following dispatch logic is a subject to be changed at any time
+
+        # Do not clean any Resubmission, but still let them be archived
+        if wflow['RequestType'] == 'Resubmission':
+            wflow['ForceArchive'] = True
+            msg = "Skipping cleanup step for workflow: %s - RequestType is %s."
+            msg += " Will try to archive it directly."
+            self.logger.info(msg, wflow['RequestName'], wflow['RequestType'])
+        # Clean:
+        elif wflow['RequestStatus'] in ['rejected', 'aborted-completed']:
+            # NOTE: We do not check the ParentageResolved flag for these
+            #       workflows, but we do need to clean output data placement
+            #       rules from the agents for them
+            for pline in self.agentlines:
+                try:
+                    pline.run(wflow)
+                except Exception as ex:
+                    msg = "%s: General error from pipeline. Workflow: %s. Error: \n%s. "
+                    msg += "\nWill retry again in the next cycle."
+                    self.logger.exception(msg, pline.name, wflow['RequestName'], str(ex))
+                    continue
+                if wflow['CleanupStatus'][pline.name]:
+                    self.wfCounters['cleaned'][pline.name] += 1
+        elif wflow['RequestStatus'] == 'announced' and not wflow['ParentageResolved']:
+            # NOTE: We skip workflows which are not having 'ParentageResolved'
+            #       flag, but we still need some proper logging for them.
+            msg = "Skipping workflow: %s - 'ParentageResolved' flag set to false."
+            msg += " Will retry again in the next cycle."
+            self.logger.info(msg, wflow['RequestName'])
+        elif wflow['RequestStatus'] == 'announced' and not wflow['TransferDone']:
+            # NOTE: We skip workflows which have not yet finalised their TransferStatus
+            #       in MSOutput, but we still need some proper logging for them.
+            msg = "Skipping workflow: %s - 'TransferStatus' is 'pending' or 'TransferInfo' is missing in MSOutput."
+            msg += " Will retry again in the next cycle."
+            self.logger.info(msg, wflow['RequestName'])
+        elif wflow['RequestStatus'] == 'announced' and wflow['ParentageResolved'] and wflow['TransferDone']:
+            for pline in self.cleanuplines:
```

Comment:
Unless I'm missing something, I don't think this code will ever be reached. We do not run the agent pipelines `self.agentlines` for workflows in `announced` status, which means we will never have `TransferDone=True`. Am I wrong?

Reply:
Sorry @amaltaro I will need you to clarify more on this, I did not get your comment here. This piece of code is executed every now and then, I do not understand two aspeccts of this comment, so I can clarify more, but :
1. What would `self.agentlines` do with the any if the conditions determining where we throw the current workflow - this variable is just a set of plines we throw at the relevant workflows nothing depends on it?
2. Why do you think we  would never have `TransferDone == True`

Reply:
Ok I think I may see your point .... but I think we may have locked out selves when we moved around some of the conditions under which we run the pipelines.

Reply:
Fixed